system/mp/mpicomm.cpp

```diff
@@ -0,0 +1,362 @@
+            {
+                mbuf.clear();
+                assertex((msg->getMessage()) != NULL);
+                mptag_t reply = msg->getMessage()->getReplyTag();
+                mbuf.transferFrom(*(msg->getMessage()));
+                mbuf.init(msg->getMessage()->getSender(),tag,reply);
+                delete msg;
+                success = true;
+            }
+        }
+        if (!messageFromSelf)
+        {
+            tm.timedout(&remaining);
+            hpcc_mpi::CommStatus status = hpcc_mpi::readData(srcrank, tag, mbuf, comm, remaining);
+            success = (status == hpcc_mpi::CommStatus::SUCCESS);
+            //TODO what if no message received and selfMsg bcomes available now for sender=RANK_ALL?
+        }
+        if (success)
+        {
+            const SocketEndpoint &ep = getGroup()->queryNode(srcrank).endpoint();
+            mbuf.init(ep, tag, mbuf.getReplyTag());
+            if (sender)
+                *sender = srcrank;
+            mbuf.reset();
+        }
+        return success;
+    }
+    
+    void barrier(void)
+    {
+        hpcc_mpi::barrier(comm);
+    }
+
+    unsigned probe(rank_t srcrank, mptag_t tag, rank_t *sender, unsigned timeout=0)
+    {
+        _TF("probe", srcrank, tag, timeout);
+        if (hpcc_mpi::hasIncomingMessage(srcrank, tag, comm))
+        {
+            if (sender)
+                *sender = srcrank;
+        }
+        return false;
+    }
+    
+    void flush(mptag_t tag)
+    {
+        // Handled by MPI
+    }
+
+    bool sendRecv(CMessageBuffer &mbuff, rank_t sendrank, mptag_t sendtag, unsigned timeout=MP_WAIT_FOREVER)
+    {
+        _TF("sendRecv", sendrank, sendtag, timeout);
+        //TODO share timeout between send/recv?
+        mptag_t replytag = createReplyTag();
+        CTimeMon tm(timeout);
+        mbuff.setReplyTag(replytag);
+        unsigned remaining;
+        if (tm.timedout(&remaining))
+            return false;
+        if (!send(mbuff,sendrank,sendtag,remaining)||tm.timedout(&remaining))
+            return false;
+        mbuff.clear();
+        return recv(mbuff,sendrank,replytag,NULL,remaining);
+    }
+
+    bool reply(CMessageBuffer &mbuf, unsigned timeout=MP_WAIT_FOREVER)
+    {
+        _TF("reply", mbuf.getReplyTag(), timeout);
+        mptag_t replytag = mbuf.getReplyTag();
+        rank_t dstrank = getGroup()->rank(mbuf.getSender());
+        if (dstrank!=RANK_NULL)
+        {
+            if (send (mbuf, dstrank, replytag,timeout))
+            {
+                mbuf.setReplyTag(TAG_NULL);
+                return true;
+            }
+            return false;
+        }
+        return false;
+    }
+
+    void cancel(rank_t srcrank, mptag_t tag)
+    {
+        _TF("cancel", srcrank, tag);
+        assertex(srcrank!=RANK_NULL);
+        //cancel only recv calls?
+        hpcc_mpi::cancelComm(false, srcrank, tag, comm);
+    }
+
+    bool verifyConnection(rank_t rank,  unsigned timeout)
+    {
+        UNIMPLEMENTED;
```

Comment:
Better to do nothing here than throw.. unless you can actually do something meaningful with MPI here.